File: 15-2.rb
#!/usr/bin/env ruby

File.read('15.input').lines.map(&:strip).each do |problem|
  seen = Hash.new([])
  numbers = problem.split(',').map(&:to_i)
  numbers.each_index do |pos|
    seen[numbers[pos]] = [pos] + seen[numbers[pos]]
  end
  numbers.size.upto (30000000 - 1) do |pos|
    if seen[numbers[pos - 1]].size == 1
      numbers.push 0
      seen[0] = [pos] + seen[0][..0]
    else
      age = seen[numbers[pos - 1]][0] - seen[numbers[pos - 1]][1]
      numbers.push age
      seen[age] = [pos] + seen[age][..0]
    end
  end
  print numbers.last, "\n"
end

File: 15-2.py
#!/usr/bin/env python

problems = [line.strip() for line in open('15.input').readlines()]

for problem in problems:
    seen = {}
    numbers = [int(n) for n in problem.split(',')]
    for pos in range(len(numbers)):
        seen[numbers[pos]] = [pos] + seen.get(numbers[pos], [])

    last = numbers[-1]
    for pos in range(len(numbers), 30000000):
        if len(seen[last]) == 1:
            last = 0
            seen[0] = [pos] + seen.get(0, [])[0:1]
        else:
            age = seen[last][0] - seen[last][1]
            last = age
            seen[age] = [pos] + seen.get(age, [])[0:1]

    print(last)
